DEVICE: amstrad cpc 464

evolutional opened the pull request:
- Adds basic `DEVICE AMSTRADCPC464` support
 - Adds `SAVESNA` support for the `AMSTRADCPC464` device

It uses the [SNA format](http://cpctech.cpc-live.com/docs/snapshot.html) popular with all CPC emulators . Have verified the built SNA on [CPCEC](http://cngsoft.no-ip.org/cpcec.htm) (Windows)

evolutional commented:
I'm planning on adding 6128 too; would you prefer it in this PR or a separate one?

evolutional commented on sjasm/io_snapshots.cpp:
```diff
@@ -145,4 +145,89 @@ int SaveSNA_ZX(char* fname, word start) {
+		0x00,		// interlace & skew
+		0x07,		// max raster
+		0x00, 0x00, // cursor start
+		0x30, 0x00, // display (xxPPSSOO) 
```
One thing that is worth discussing / deciding here is if we want to move SAVESNA into a separate directive, such as OPENCPCSNA and allow customization of some of these defaults; notably the PEN settings and display address may be something a user wants to configure.

I figured implement the simplest thing and change if we think it's needed

ped7g replied:
from the used DEVICE it is clear which `SAVESNA` implementation has to be used, so in this regard I guess the end user doesn't really care about the name collision and the CPC `SAVESNA` may be custom-tailored as much as you want.

But thinking about the "searching documentation" use case, maybe it may be worth to give it custom mnemonics too, so users will not read ZX SAVESNA docs by accidents... about the open/close way like NEX file has vs single SAVESNA (or CPCSAVE or something like that)... the single directive is often easier to use, often outweighing the details, for example ZX SNA can't be configured to have interrupts enabled or disable (although it's trivial to post-patch the outputted file directly form asm source, as it's not encrypted or compressed). 99% of ZX snapshots are fine with defaults in those extra fields. If you know CPC developers like to customise snapshots more, it may be worth to do some open->close stuff like NEX, including full config.
(but generally on ZX the snapshot format is kinda frown upon, it's nice for quick-hack development, but final release should be rather proper TAP image, just dumping whole machine into snap is considered a bit lame... no idea how CPC scene feels about that... if it's similar, then I guess fine configuring of quick-hack format is not very valuable).

evolutional replied:
Yeah, I agree. 

The snapshot format has similar principles; except it's safer on the CPC emulators (no stack modification). Most people distribute stuff as TZX / CDT format (CDT is just TZX by another name). SNA here is used in the same way; eg: hack a build together for dev/testing. I agree though, I'll move the directive to SAVECPC or similar to avoid confusion - will have to have a think about if we care about having all the tweakable values or not. I know I don't need them as I set them on program init :)

I'm also working on adding some simple TZX support for both Speccy and CPC and that follows the SAVENEX style directives.

evolutional commented:
> cram this into 1.18.3

No rush. I added it because I wanted use it for my personal development (sjasmplus is my favourite assembler); figured I'd start a PR in case anyone else is interested.

evolutional commented:
Have cleaned up the original PR; renamed to `SAVECPCSNA`. I've also moved the CPC SNA code into a CPC file, mostly because I want to add tape support in the future and it's best to have all the CPC stuff together.

Still todo:
 - add CPC 6128 support
 - decide if we need the additional directives to control paging and the crtc/pens.
 
On the latter, I was talking to a CPC dev with more experience and he thinks that we won't really need the additional directives, as most people will set this up in their code (they'll need it for tape loading, for example).

evolutional commented on sjasm/directives.cpp:
```diff
@@ -702,8 +702,14 @@ static void dirSAVESNA() {
+		} else if (IsAmstradCPCDevice(DeviceID)) {
+			exec = SaveSNA_CPC(fnaam.get(), start);
+		}
+		if (!exec)
```
Done, moved this out into its own directive.

evolutional commented on sjasm/io_snapshots.cpp:
```diff
@@ -145,4 +145,89 @@ int SaveSNA_ZX(char* fname, word start) {
+	FILE* ff;
+	if (!FOPEN_ISOK(ff, fname, "wb")) {
+		Error("[SAVESNA] Error opening file", fname, FATAL);
+		return 0;
```
Done!

evolutional commented on sjasm/io_snapshots.cpp:
```diff
@@ -145,4 +145,89 @@ int SaveSNA_ZX(char* fname, word start) {
+	// v1 format fields
+	snbuf[0x1B] = 0; snbuf[0x1C] = 0; // ensure interrupts are disabled
+	snbuf[0x21] = 0xF0; snbuf[0x22] = 0xBF; // sp to BFF0
+	snbuf[0x23] = char(start & 0xFF); snbuf[0x24] = char(start >> 8); // pc
```
Done!

ped7g commented:
...merging...

ped7g commented:
merge done (with minor changes), closing PR.
